clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesV2OnDemandCachingAgent.java

```diff
@@ -0,0 +1,309 @@
+  }
+
+  private void processOnDemandEntry(CacheData onDemandEntry) {
+    Map<String, Object> attributes = onDemandEntry.getAttributes();
+    Integer processedCount = (Integer) attributes.get(PROCESSED_COUNT_KEY);
+    Long processedTime = System.currentTimeMillis();
+
+    processedCount = processedCount == null ? 0 : processedCount;
+    processedCount += 1;
+
+    attributes.put(PROCESSED_TIME_KEY, processedTime);
+    attributes.put(PROCESSED_COUNT_KEY, processedCount);
+  }
+
+  private boolean shouldKeepInOnDemand(Long lastFullRefresh, CacheData onDemandEntry) {
+    Map<String, Object> attributes = onDemandEntry.getAttributes();
+    Long cacheTime = (Long) attributes.get(CACHE_TIME_KEY);
+    Integer processedCount = (Integer) attributes.get(PROCESSED_COUNT_KEY);
+
+    cacheTime = cacheTime == null ? 0L : cacheTime;
+    processedCount = processedCount == null ? 0 : processedCount;
+
+    return cacheTime >= lastFullRefresh || processedCount == 0;
+  }
+
+  private OnDemandAgent.OnDemandResult evictEntry(ProviderCache providerCache, String key) {
+    Map<String, Collection<String>> evictions = new HashMap<>();
+    CacheResult cacheResult = new DefaultCacheResult(new HashMap<>());
+
+    log.info("Evicting on demand '{}'", key);
+    providerCache.evictDeletedItems(ON_DEMAND_TYPE, Collections.singletonList(key));
+    evictions.put(primaryKind().toString(), Collections.singletonList(key));
+
+    return new OnDemandAgent.OnDemandResult(getOnDemandAgentType(), cacheResult, evictions);
+  }
+
+  private OnDemandAgent.OnDemandResult addEntry(ProviderCache providerCache, String key, T resource) throws JsonProcessingException {
+    Map<String, Collection<String>> evictions = new HashMap<>();
+    CacheResult cacheResult;
+
+    log.info("Storing on demand '{}'", key);
+    cacheResult = buildCacheResult(resource);
+    String jsonResult = objectMapper.writeValueAsString(cacheResult.getCacheResults());
+
+    Map<String, Object> attributes = new ImmutableMap.Builder<String, Object>()
+        .put(CACHE_TIME_KEY, System.currentTimeMillis())
+        .put(CACHE_RESULTS_KEY, jsonResult)
+        .put(PROCESSED_COUNT_KEY, 0)
+        .put(PROCESSED_TIME_KEY, null)
+        .build();
+
+    Map<String, Collection<String>> relationships = new HashMap<>();
+    CacheData onDemandData = new DefaultCacheData(key, attributes, relationships);
+    providerCache.putCacheData(ON_DEMAND_TYPE, onDemandData);
+
+    return new OnDemandAgent.OnDemandResult(getOnDemandAgentType(), cacheResult, evictions);
+  }
+
+  @Override
+  public OnDemandAgent.OnDemandResult handle(ProviderCache providerCache, Map<String, ?> data) {
+    String account = (String) data.get("accountName");
+    String name = getResourceNameFromOnDemandRequest(data).orElse("");
+    String namespace = (String) data.get("namespace");
+    if (StringUtils.isEmpty(namespace)) {
+      namespace = (String) data.get("region"); // sigh
```

Comment:
Maybe provide details for the `sigh` or leave it out. Seems like there is a helper function that could help self document why you have to `(String) data.get()` everything.